src/utils/dom-stream.js

```diff
@@ -0,0 +1,200 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+const MAX_DEPTH = 10;
+const FORBIDDEN_SPLIT = {
+  'TABLE': 1,
+  'UL': 1,
+  'OL': 1,
+  'DL': 1,
+};
+
+
+/**
+ * Stream-merges the input DOM into the target. The `flush()` method can be
+ * called as often as needed, but it's recommended to only call it in vsync
+ * mutate.
+ */
+export class DomStream {
+
+  /**
+   * @param {!Node} input
+   * @param {!Node} target
+   */
+  constructor(input, target) {
+    /** @const @private {!Node} */
+    this.input_ = input;
+
+    /** @private {?Node} */
+    this.inputPoint_ = null;
+
+    /** @const @private {!Node} */
+    this.target_ = target;
+
+    /** @private {!Node} */
+    this.targetPoint_ = target;
+
+    /** @const @private {number} */
+    this.maxDepth_ = MAX_DEPTH;
+
+    /** @private {number} */
+    this.depth_ = 0;
+  }
+
+  /**
+   * Merges all confirmed completed nodes from input to target. It's recommended
+   * to always call this method in vsync mutate phase.
+   */
+  flush() {
+    this.flush_(/* eof */ false);
+  }
+
+  /**
+   * Signals that the input DOM is complete. Merges all pending nodes into
+   * the target DOM. It's recommended to always call this method in vsync
+   * mutate phase.
+   */
+  done() {
+    this.flush_(/* eof */ true);
+  }
+
+  /**
+   * @param {boolean} eof
+   * @private
+   */
+  flush_(eof) {
+    // There are two phases:
+    // 1. "advance". This phase waits for a sibling node to the pending node
+    //    or a singling node to one of its ancestors. As soon as sibling node
+    //    is found, the parse of pending node is complete and it can be merged
+    //    into the target DOM.
+    // 2. "split". If no sibling nodes are found, this phase finds the nearest
+    //    descendant of this node with a sibling. If found, it imports the
+    //    partial DOM path and the found descendant node into the target DOM.
+    //    There are limitations of how deep the input DOM can be split and some
+    //    nodes are not allowed to be split, such as "table" and "ul".
```

Comment:
Would you mind adding HTML examples to this comment as a sort of timeline for when things happen?